crates/wac-graph/src/plug.rs

```diff
@@ -0,0 +1,64 @@
+use crate::{
+    types::{Package, SubtypeChecker},
+    CompositionGraph, EncodeOptions, NodeId, PackageId,
+};
+
+/// Take the exports of the plug component and plug them into the socket component.
+pub fn plug(
+    graph: &mut CompositionGraph,
+    plugs: Vec<Package>,
+    socket: PackageId,
+    socket_instantiation: NodeId,
+) -> Result<Vec<u8>, anyhow::Error> {
+    for plug_package in plugs {
+        let plug = graph.register_package(plug_package)?;
+
+        let mut plugs = Vec::new();
+        let mut cache = Default::default();
+        let mut checker = SubtypeChecker::new(&mut cache);
+        for (name, plug_ty) in &graph.types()[graph[plug].ty()].exports {
+            if let Some(socket_ty) = graph.types()[graph[socket].ty()].imports.get(name) {
+                if checker
+                    .is_subtype(*plug_ty, graph.types(), *socket_ty, graph.types())
+                    .is_ok()
+                {
+                    plugs.push(name.clone());
+                }
+            }
+        }
+
+        // Instantiate the plug component
+        let mut plug_instantiation = None;
+        for plug_name in plugs {
+            log::debug!("using export `{plug_name}` for plug");
+            let plug_instantiation =
+                *plug_instantiation.get_or_insert_with(|| graph.instantiate(plug));
+            let export = graph.alias_instance_export(plug_instantiation, &plug_name)?;
+            graph.set_instantiation_argument(socket_instantiation, &plug_name, export)?;
+        }
+    }
+
+    // Check we've actually done any plugging.
+    if graph
+        .get_instantiation_arguments(socket_instantiation)
+        .next()
+        .is_none()
+    {
+        anyhow::bail!(
+            "the socket component had no matching imports for the plugs that were provided"
+        )
+    }
+
+    // Export all exports from the socket component.
+    for name in graph.types()[graph[socket].ty()]
+        .exports
+        .keys()
+        .cloned()
+        .collect::<Vec<_>>()
+    {
+        let export = graph.alias_instance_export(socket_instantiation, &name)?;
+        graph.export(export, &name)?;
+    }
+
+    Ok(graph.encode(EncodeOptions::default())?)
```

Comment:
I think we should leave the encoding to the caller. So this function will just return `Ok(())` and if the caller is ready to encode, they can do the call to `graph.encode`.